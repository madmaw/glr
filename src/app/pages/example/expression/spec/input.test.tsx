import {
  ExpressionInput,
  MSG_BUTTON_TEXT,
  MSG_INPUT_PLACEHOLDER,
} from 'app/pages/example/expression/input';
import {
  createEvent,
  fireEvent,
  render,
} from 'app/pages/example/testing/react';
import { type Mock } from 'vitest';

suite('ExpressionInput', function () {
  let onChangeExpression: Mock<[string], void>;
  let onEvaluateExpression: Mock<[], void>;

  beforeEach(function () {
    onChangeExpression = vitest.fn();
    onEvaluateExpression = vitest.fn();
  });

  describe.each([
    '123',
    '"abc"',
    '',
    '  return',
  ])('with expression %s', function (expression) {
    describe.each([
      true,
      false,
    ])('disabled %s', function (disabled) {
      it('renders', function () {
        const { container } = render(
          (
            <ExpressionInput
              disabled={disabled}
              expression={expression}
              onChangeExpression={onChangeExpression}
              onEvaluateExpression={onEvaluateExpression}
            />
          ),
        );
        expect(container).toMatchSnapshot();
      });
    });
  });

  it('fires on change event', function () {
    const newValue = '2';
    const { getByPlaceholderText } = render(
      (
        <ExpressionInput
          disabled={false}
          expression='1'
          onChangeExpression={onChangeExpression}
          onEvaluateExpression={onEvaluateExpression}
        />
      ),
    );
    const input = getByPlaceholderText(MSG_INPUT_PLACEHOLDER());
    expect(input).toBeDefined();
    expect(onChangeExpression).not.toHaveBeenCalled();
    const event = createEvent.change(input, { target: { value: newValue } });
    fireEvent(input, event);
    expect(onChangeExpression).toHaveBeenCalledOnce();
    expect(onChangeExpression).toHaveBeenCalledWith(newValue);
  });

  it('fires evaluate expression event from button', function () {
    const { getByText } = render(
      (
        <ExpressionInput
          disabled={false}
          expression='1'
          onChangeExpression={onChangeExpression}
          onEvaluateExpression={onEvaluateExpression}
        />
      ),
    );
    const button = getByText(MSG_BUTTON_TEXT());
    expect(button).toBeDefined();
    expect(onEvaluateExpression).not.toHaveBeenCalled();
    button.click();
    expect(onEvaluateExpression).toHaveBeenCalledOnce();
  });

  it('fires evaluate expression event from input on enter key', function () {
    // TODO
  });
});
