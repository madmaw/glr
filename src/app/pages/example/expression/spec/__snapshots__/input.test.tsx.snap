// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExpressionInput > with expression   return > disabled false > renders 1`] = `
<div>
  <div
    class="css-995cds"
    direction="row"
  >
    <input
      class="css-1rrnh34-bordered-typographic"
      placeholder="example: 1 + 1"
      value="  return"
    />
    <button
      class="css-s6v2zt-typographic-bordered"
    >
      Evaluate!
    </button>
  </div>
</div>
`;

exports[`ExpressionInput > with expression   return > disabled true > renders 1`] = `
<div>
  <div
    class="css-995cds"
    direction="row"
  >
    <input
      class="css-1rrnh34-bordered-typographic"
      disabled=""
      placeholder="example: 1 + 1"
      value="  return"
    />
    <button
      class="css-s6v2zt-typographic-bordered"
    >
      Evaluate!
    </button>
  </div>
</div>
`;

exports[`ExpressionInput > with expression  > disabled false > renders 1`] = `
<div>
  <div
    class="css-995cds"
    direction="row"
  >
    <input
      class="css-1rrnh34-bordered-typographic"
      placeholder="example: 1 + 1"
      value=""
    />
    <button
      class="css-s6v2zt-typographic-bordered"
    >
      Evaluate!
    </button>
  </div>
</div>
`;

exports[`ExpressionInput > with expression  > disabled true > renders 1`] = `
<div>
  <div
    class="css-995cds"
    direction="row"
  >
    <input
      class="css-1rrnh34-bordered-typographic"
      disabled=""
      placeholder="example: 1 + 1"
      value=""
    />
    <button
      class="css-s6v2zt-typographic-bordered"
    >
      Evaluate!
    </button>
  </div>
</div>
`;

exports[`ExpressionInput > with expression "abc" > disabled false > renders 1`] = `
<div>
  <div
    class="css-995cds"
    direction="row"
  >
    <input
      class="css-1rrnh34-bordered-typographic"
      placeholder="example: 1 + 1"
      value=""abc""
    />
    <button
      class="css-s6v2zt-typographic-bordered"
    >
      Evaluate!
    </button>
  </div>
</div>
`;

exports[`ExpressionInput > with expression "abc" > disabled true > renders 1`] = `
<div>
  <div
    class="css-995cds"
    direction="row"
  >
    <input
      class="css-1rrnh34-bordered-typographic"
      disabled=""
      placeholder="example: 1 + 1"
      value=""abc""
    />
    <button
      class="css-s6v2zt-typographic-bordered"
    >
      Evaluate!
    </button>
  </div>
</div>
`;

exports[`ExpressionInput > with expression 123 > disabled false > renders 1`] = `
<div>
  <div
    class="css-995cds"
    direction="row"
  >
    <input
      class="css-1rrnh34-bordered-typographic"
      placeholder="example: 1 + 1"
      value="123"
    />
    <button
      class="css-s6v2zt-typographic-bordered"
    >
      Evaluate!
    </button>
  </div>
</div>
`;

exports[`ExpressionInput > with expression 123 > disabled true > renders 1`] = `
<div>
  <div
    class="css-995cds"
    direction="row"
  >
    <input
      class="css-1rrnh34-bordered-typographic"
      disabled=""
      placeholder="example: 1 + 1"
      value="123"
    />
    <button
      class="css-s6v2zt-typographic-bordered"
    >
      Evaluate!
    </button>
  </div>
</div>
`;
